### PR TITLE
Implement ``align``

### PR DESCRIPTION
**Pull request by phofl**
Not super happy about the structure here but should be ok for a first attempt

### CODE REVIEW

File: dask_expr/collection.py [hunk 1]
```diff
@@ -44,6 +44,10 @@ def _wrap_expr_api(*args, wrap_api=None, **kwargs):
     )
     if isinstance(result, expr.Expr):
         return new_collection(result)
+    elif isinstance(result, tuple):
+        return tuple(
```
Comment (phofl) [inline dask_expr/collection.py:48]: We could also do this in ``align`` itself

File: dask_expr/expr.py [hunk 2]
```diff
@@ -871,6 +875,29 @@ class Replace(Blockwise):
     operation = M.replace
 
 
+class _Align(Blockwise):
+    _parameters = ["frame", "other", "join", "fill_value"]
+    _defaults = {"join": "outer", "fill_value": None}
+    _keyword_only = ["join", "fill_value"]
+    operation = M.align
```
Comment (mrocklin) [inline dask_expr/expr.py:882]: Is this truly blockwise though?  What if one index values is on one partition in one dataframe, but in another partition in the other dataframe?

Comment (phofl) [reply dask_expr/expr.py:884]: I forgot adding a Repartition call before this class. The whole operation isn't blockwise, but this class is

File: dask_expr/expr.py [hunk 3]
```diff
@@ -439,6 +440,30 @@ def apply(self, function, *args, **kwargs):
...
+                "partitions. Please use `set_index` "
+                "to set the index."
+            )
+
+        divisions = list(unique(merge_sorted(*[df.divisions for df in dfs])))
+        if len(divisions) == 1:  # single value for index
+            divisions = (divisions[0], divisions[0])
+
+        left = Repartition(self, new_divisions=divisions, force=True)
+        other = Repartition(other, new_divisions=divisions, force=True)
```
Comment (mrocklin) [inline dask_expr/expr.py:460]: If the divisions are the same then this should probably be a no-op.  That logic should either be in the `.repartition` method or in a `simplify_down` method of the `Repartition` class.

File: demo.ipynb [hunk 4]
```diff
@@ -18,12 +18,23 @@
   },
   {
    "cell_type": "code",
-   "execution_count": null,
+   "execution_count": 39,
```
Comment (mrocklin) [inline demo.ipynb:21]: It looks like the notebook was modified for this PR.  Maybe an errant `git commit -a`

File: dask_expr/expr.py [hunk 5]
```diff
@@ -439,6 +440,30 @@ def apply(self, function, *args, **kwargs):
...
+        aligned = _Align(left, other, join=join, fill_value=fill_value)
```
Comment (mrocklin) [inline dask_expr/expr.py:461]: Interesting, so this intermediate object isn't a dataframe, it's a tuple of two dataframes?  Totally fine, just interesting.  It breaks our current model a little (but in a way that's interesting to think about).

File: dask_expr/expr.py (hunk 3 above)
Comment (mrocklin) [reply dask_expr/expr.py:460]: Happy if this is future work though.

**Review by mrocklin — APPROVED** [review]

File: dask_expr/expr.py (hunk 3 above)
Comment (phofl) [reply dask_expr/expr.py:460]: I'll put it on my list

File: dask_expr/expr.py (hunk 5 above)
Comment (phofl) [reply dask_expr/expr.py:461]: Yes exactly. I think this is the cleanest in the sense that we don't deviate from the pandas model. But we could certainly change this at some point to merge both classes together

**Comment by phofl** [comment]
Addressed the repartition suggestion in #162, so this is ready

**Pull request merged**